### PR TITLE
Introduce quarkus.quinoa.package-manager-command

### PR DESCRIPTION
**Pull request by ia3andy**

### CODE REVIEW

**Comment by ia3andy** [comment]
That looks pretty good, a few comments:
- How can we deal with prod/dev builds when using `package-manager-commands`? shouldn't we propage the "built-in" headers
- I would name it `package-manager-command` (no s) example of similar case in `quarkus.http.header` https://quarkus.io/guides/all-config#quarkus-vertx-http_quarkus.http.header.-header-.path
- Now that's it's getting big, we should put this all the package manager logic in a new java package and split the class in multiple
- The `binary()` overriding logic is getting hard to follow (maybe that's just me I need to have a closer look)
- How will the frozen-lockfile config works with this? 

For the envs, I was initialy thinking about making them part of the command itself (e.g. `FOO=BAR; npm intstall`) but maybe your solution is better/safer (cc @maxandersen)

File: deployment/src/main/java/io/quarkiverse/quinoa/deployment/PackageManager.java [hunk 1]
```diff
@@ -225,61 +215,138 @@ private Command(Map<String, String> envs, String... args) {
...
+            return new Command(
+                    commandsConfig.devEnv.isEmpty() ? c.envs : commandsConfig.devEnv,
+                    mapToArray(commandsConfig.dev).orElse(c.args));
+        }
+
+        private Optional<String[]> mapToArray(Optional<String> command) {
+            return command.map(new Function<String, String[]>() {
+                @Override
+                public String[] apply(String s) {
+                    return s.split(" ");
```
Comment (ia3andy) [inline deployment/src/main/java/io/quarkiverse/quinoa/deployment/PackageManager.java:347]:
```suggestion
                    return s.split("\\s+");
```

Something like this might me more flexible

**Comment by ia3andy** [comment]
@computerlove have you seen my comments? 👆

File: deployment/src/main/java/io/quarkiverse/quinoa/deployment/PackageManager.java (hunk 1 above)
Comment (ia3andy) [reply deployment/src/main/java/io/quarkiverse/quinoa/deployment/PackageManager.java:359]: Maybe this can help: https://stackoverflow.com/questions/1757065/java-splitting-a-comma-separated-string-but-ignoring-commas-in-quotes

**Comment by computerlove** [comment]
I'm leaning towards changing command handling to use only a single string internally, and sending this one string to `commons-exec` rather than trying to get the splitting right.

**Comment by ia3andy** [comment]
> I'm leaning towards changing command handling to use only a single string internally, and sending this one string to `commons-exec` rather than trying to get the splitting right.

if you do that you could also rmove the `-env`, maybe it's worth a try?

**Comment by ia3andy** [comment]
> I'm leaning towards changing command handling to use only a single string internally, and sending this one string to `commons-exec` rather than trying to get the splitting right.

Without calling using `commons-exec` you can use `/bin/sh -c` and pass the entire command string (`cmd /c` on Windows)

See https://www.baeldung.com/run-shell-command-in-java

**Comment by ia3andy** [comment]
@computerlove let me know when you feel it should be reviewed :)

**Comment by ia3andy** [comment]
> Should I change `public final String[] args;` to just `String`, do you think? I'll update the documentation.
> 
> Other than that I don't think there is more to be done :)

It's up to you :)

I also thought about it while reading:
👍 Custom and Automatic would be more similar
👎 For Automatic, concatenating strings is not as nice as providing an array
 
For the env configs, we also have the option to remove them:
- It would significantly reduce the qty of configuration options
- You could still use the command string to provide the envs (`FOO=BAR; npm install`)
- But, having those as configs is maybe better for cross-platform compat

File: deployment/src/main/java/io/quarkiverse/quinoa/deployment/packagemanager/Commands.java [hunk 2]
```diff
@@ -0,0 +1,26 @@
+package io.quarkiverse.quinoa.deployment.packagemanager;
+
+import java.util.Collections;
+
+import io.quarkus.runtime.LaunchMode;
+
+interface Commands {
```
Comment (ia3andy) [inline deployment/src/main/java/io/quarkiverse/quinoa/deployment/packagemanager/Commands.java:7]: Maybe `PackageManagerCommands` or rename `Command` to `ExecCommand` (to make it a bit more readable). WDYT?

**Review by ia3andy — COMMENTED** [review]
Really nice PR :) Thanks!!!

I just added a few naming suggestions (you are free to ignore them) + a missing equals/hashCode

**Comment by ia3andy** [comment]
@all-contributors add @computerlove for code

**Comment by ia3andy** [comment]
I will be back on monday, everything's seems perfect I will just give it a try before merging 👍
